### ISSUE TITLE
[QUESTION]: is it possible to detect errors in kernel operations?

### ISSUE THREAD

**Issue by El-Gor-do** [m4rs-mt/ILGPU#1305]
### Question

If a kernel operation fails but doesn't kill my application, is it possible to detect that an error has occurred? In my sample code below, I have 2 test kernels - `DivideByZeroKernel` and `IndexOutOfRangeKernel`. 

```
public static void Run()
{
    int[] inputData = new int[] { 0, -100 };

    using (Context context = Context.CreateDefault())
    using (Accelerator acc = context.CreateCudaAccelerator(0))
    using (MemoryBuffer1D<int, Stride1D.Dense> deviceData = acc.Allocate1D(inputData))
    using (MemoryBuffer1D<int, Stride1D.Dense> deviceOutput = acc.Allocate1D<int>(deviceData.Length))
    {
        Action<Index1D, ArrayView<int>, ArrayView<int>> kernel = DivideByZeroKernel;    // DivideByZeroKernel or IndexOutOfRangeKernel
        Action<Index1D, ArrayView<int>, ArrayView<int>> loadedKernel = acc.LoadAutoGroupedStreamKernel<Index1D, ArrayView<int>, ArrayView<int>>(kernel);
        loadedKernel((int)deviceOutput.Length, deviceData.View, deviceOutput.View);
        try
        {
            acc.Synchronize();

            int[] hostOutput = deviceOutput.GetAsArray1D();
            for (int i = 0; i < hostOutput.Length; ++i)
            {
                Console.WriteLine($"100 / {inputData[i]} = {hostOutput[i]}");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.ToString());
        }
    }   // using IndexOutOfRangeKernel, Visual Studio 2022 breaks on this line with ILGPU.Runtime.Cuda.CudaException: 'device-side assert triggered'
}

private static void DivideByZeroKernel(Index1D i, ArrayView<int> data, ArrayView<int> output)
{
    output[i] = 100 / data[i];
}

private static void IndexOutOfRangeKernel(Index1D i, ArrayView<int> data, ArrayView<int> output)
{
    output[i] = 100 / data[i + 1];
}
```

The output when running `DivideByZeroKernel` is below. Even though I divide by zero, the operation returned -1 instead of infinity, NaN or crashing. Is there a way to detect that -1 is an error value in the first case but a valid result in the second case? 

```
100 / 0 = -1
100 / -100 = -1
```

The output when running `IndexOutOfRangeKernel` is below. Based on https://github.com/m4rs-mt/ILGPU/issues/1268 it would appear that my application will crash even if `acc.Synchronize()` is wrapped in a `try/catch`. I could use a sidecar app to monitor if my GPU app has crashed - are there any other ways to detect that a GPU kernel has crashed my app?
```
C:\temp\IlgpuApp\GpuErrors.cs:46: block: [0,0,0], thread: [1,0,0] Assertion `Index out of range` failed.
ILGPU.Runtime.Cuda.CudaException: device-side assert triggered
   at ILGPU.Runtime.Cuda.CudaAccelerator.SynchronizeInternal()
   at ILGPU.Runtime.Accelerator.Synchronize()
   at IlgpuApp.GpuErrors.Run() in C:\temp\IlgpuApp\GpuErrors.cs:line 24
Unhandled exception. ILGPU.Runtime.Cuda.CudaException: device-side assert triggered
   at ILGPU.Runtime.Cuda.CudaMemoryBuffer.DisposeAcceleratorObject(Boolean disposing)
   at ILGPU.Runtime.AcceleratorObject.DisposeAcceleratorObject_Accelerator(Boolean disposing)
   at ILGPU.Runtime.Accelerator.DisposeChildObject_AcceleratorObject(AcceleratorObject acceleratorObject, Boolean disposing)
   at ILGPU.Runtime.AcceleratorObject.Dispose(Boolean disposing)
   at ILGPU.Util.DisposeBase.DisposeDriver(Boolean disposing)
   at ILGPU.Util.DisposeBase.Dispose()
   at ILGPU.Runtime.MemoryBuffer`1.DisposeAcceleratorObject(Boolean disposing)
   at ILGPU.Runtime.AcceleratorObject.DisposeAcceleratorObject_Accelerator(Boolean disposing)
   at ILGPU.Runtime.Accelerator.DisposeChildObject_AcceleratorObject(AcceleratorObject acceleratorObject, Boolean disposing)
   at ILGPU.Runtime.AcceleratorObject.Dispose(Boolean disposing)
   at ILGPU.Util.DisposeBase.DisposeDriver(Boolean disposing)
   at ILGPU.Util.DisposeBase.Dispose()
   at IlgpuApp.GpuErrors.Run() in C:\temp\IlgpuApp\GpuErrors.cs:line 36
   at IlgpuApp.Program.Main(String[] args) in C:\temp\IlgpuApp\Program.cs:line 14
```

### Environment

- ILGPU version: 1.5.1
- .NET version: .NET 8
- Operating system: Windows 11
- Hardware (if GPU-related): NVIDIA GeForce GTX 3060

**Comment by MoFtZ** [comment]
@El-Gor-do ok, did some more investigation, and it looks like Cuda will not throw an exception on divide-by-zero. This is consistent with the [exception handling rules](https://en.wikipedia.org/wiki/IEEE_754#Exception_handling) for the IEEE 754 standard. It looks like it will return +infinity or -infinity.

Looking at other Cuda libraries, they have written their own code to detect divide-by-zero and abort the kernel.

**Comment by El-Gor-do** [comment]
If IEEE754 says divide by zero should be +/- infinity, is there a bug in ILGPU? Divide by zero is returning -1 which is why I've been asking how I can detect that an error has occurred.

**Comment by MoFtZ** [comment]
I have checked the behavior of divide by zero, and it is giving me infinity. Looking back through your example, you are using integers not floating points. Since IEEE 754 only applies to floating point operations, the +/- infinity result does not apply.

I checked the behavior of divide by zero on integers, and yes, I am getting -1 too (Cuda 12 SDK, 1070 GPU). My guess is that you cannot rely on this -1 result, since divide by zero on integers is undefined. Cuda can change their behavior at any time. It just so happens that .NET throws an exception.

So in summary, no, there is no way to automatically check for divide by zero errors. You would need to write code to check the denominator yourself.

**Comment by El-Gor-do** [comment]
After an unrecoverable error, e.g. using `IndexOutOfRangeKernel`, what is the correct pattern for cleaning up `IDisposable`s? In my test code, `deviceOutput`, `deviceData`, `acc` and `context` are all `IDisposable` but attempting to dispose `deviceOutput`, `deviceData` or `acc` will throw `Unhandled exception. ILGPU.Runtime.Cuda.CudaException: device-side assert triggered`. Only `context` successfully disposes. Do I even need to dispose `deviceOutput`, `deviceData` and `acc` or is it sufficient to dispose `context`?

**Comment by MoFtZ** [comment]
@El-Gor-do Cuda is a native library. It is important  to cleanup all the resources you acquire. ILGPU includes Finalizers that will attempt to cleanup the native resources, however, the order of cleanup is non-deterministic.

**Comment by El-Gor-do** [comment]
Yes I am aware that I need to clean up `IDisposable` resources - that's what I'm doing in the `finally` block. My question is whether this is the correct way to clean up resources after an unrecoverable error because `.Dispose()` is throwing `ILGPU.Runtime.Cuda.CudaException`

**Comment by MoFtZ** [comment]
The exception is an assertion failure in the Cuda kernel. You should be fixing your kernel to avoid triggering the assertion. In release mode, assertions are normally excluded.
